src/mapeditor.c

```diff
@@ -826,183 +824,187 @@ static void confirm_delete(void *ptr)
-		.destsurf = dd->wndsurf,
-		.center = { dd->wndsurf->w/2 + button_width(0)/2, dd->wndsurf->h/2 },
+		.destsurf = ed->cam.surface,
+		.center = { ed->cam.surface->w/2 + button_width(0)/2, ed->cam.surface->h/2 },
 		.onclick = set_to_true,
 		.onclickdata = &noclicked,
 	};
 
 	button_show(&yesbtn);
 	button_show(&nobtn);
-	misc_blit_with_center(textsurf, dd->wndsurf, &(SDL_Point){ dd->wndsurf->w/2, dd->wndsurf->h/4 });
+	misc_blit_with_center(textsurf, ed->cam.surface, &(SDL_Point){ ed->cam.surface->w/2, ed->cam.surface->h/4 });
 
 	struct LoopTimer lt = {0};
 	while(!yesclicked && !noclicked) {
 		SDL_Event e;
 		while (SDL_PollEvent(&e)) {
 			if (e.type == SDL_QUIT) {
-				dd->editor->state = MISC_STATE_QUIT;
+				ed->state = MISC_STATE_QUIT;
 				goto out;
 			}
 			button_handle_event(&e, &yesbtn);
 			button_handle_event(&e, &nobtn);
 		}
-		SDL_UpdateWindowSurface(dd->wnd);
+		SDL_assert(ed->wnd);
+		SDL_UpdateWindowSurface(ed->wnd);
 		looptimer_wait(&lt);
 	}
 
 	if (yesclicked) {
-		map_delete(dd->maps, dd->nmaps, dd->editor->map - dd->maps);
-		dd->editor->state = MISC_STATE_CHOOSER;
+		map_delete(ed->maps, ed->nmaps, ed->map - ed->maps);
+		ed->state = MISC_STATE_CHOOSER;
 	}
 
 out:
 	SDL_FreeSurface(textsurf);
 }
 
-enum MiscState mapeditor_run(
-	SDL_Window *wnd,
-	struct Map *maps, int *nmaps, int mapidx,
-	const struct EllipsoidPic *plr0pic, const struct EllipsoidPic *plr1pic)
+void mapeditor_reinit(struct MapEditor *ed, struct Map *maps, int *nmaps, int mapidx)
 {
-	struct Map *map = &maps[mapidx];
-
-	SDL_Surface *wndsurf = SDL_GetWindowSurface(wnd);
-	if (!wndsurf)
-		log_printf_abort("SDL_GetWindowSurface failed: %s", SDL_GetError());
-	SDL_FillRect(wndsurf, NULL, 0);
-
-	struct DeleteData deldata = {
-		.wnd = wnd,
-		.wndsurf = wndsurf,
-		.maps = maps,
-		.nmaps = nmaps,
-	};
-	struct MapEditor ed = {
-		.sel = { .mode = SEL_NONE },
-		.state = MISC_STATE_MAPEDITOR,
-		.map = map,
-		.playeredits = {
-			{
-				.el = {
-					.xzradius = PLAYER_XZRADIUS,
-					.yradius = PLAYER_YRADIUS_NOFLAT,
-					.epic = plr0pic,
-					.center = { .y = PLAYER_YRADIUS_NOFLAT },
-				},
-				.loc = &map->playerlocs[0],
-			},
-			{
-				.el = {
-					.xzradius = PLAYER_XZRADIUS,
-					.yradius = PLAYER_YRADIUS_NOFLAT,
-					.epic = plr1pic,
-					.center = { .y = PLAYER_YRADIUS_NOFLAT },
-				},
-				.loc = &map->playerlocs[1],
-			},
-		},
-		.cam = {
-			.screencentery = 0,
-			.surface = misc_create_cropped_surface(wndsurf, (SDL_Rect){
-				0, 0,
-				CAMERA_SCREEN_WIDTH, CAMERA_SCREEN_HEIGHT,
-			}),
-			.angle = 0,
-		},
-		.rotatedir = 0,
-		.donebtn = {
-			.text = "Done",
-			.destsurf = wndsurf,
-			.center = {
-				button_width(0)/2,
-				button_height(0)/2
-			},
-			.scancodes = { SDL_SCANCODE_ESCAPE },
-			.onclick = on_done_clicked,
-			.onclickdata = &ed,
+	SDL_FillRect(ed->cam.surface, NULL, 0);
+
+	ed->map = &maps[mapidx];
+	ed->maps = maps;
+	ed->nmaps = nmaps;
+	ed->redraw = true;   // first iteration always redraws
```

Comment:
Now that this is here instead of a weird for loop trick, it's obvious enough to not require comment

```suggestion
	ed->redraw = true;
```